protocol/network/4-architecture.md

```diff
@@ -0,0 +1,126 @@
+4 Architecture
+==============
+
+libp2p was designed around the Unix Philosophy, creating smaller components, easier to understand and to test. These components should also be able to be swapped in order to accomodate different technologies or scenarios and also make it that it is upgradable over time.
+
+Although different Peers can support different protocols depending on their capabilities, any Peer can act as a dialer and/or a listener for connections from other Peers, connections that once established can be reused from both ends, removing the distinction between clients and servers.
+
+libp2p interface acts as a thin veneer to three subsystems that are required in order for peers to be able to communicate. These subsystems are allowed to be built on top of other subsystems as long as they respect the standardized interface. The main 3 subsystems are:
+
+- Peer Routing - Mechanism to find a Peer in a network. This Routing can be done recursively, iteratively or even in a broadcast/multicast mode.
+- Swarm - Handles everything that touches the 'opening a stream' part of libp2p, from protocol muxing, stream muxing, NAT Traversal, Connection Relaying, while being multitransport
```

Comment:
> @diasdavid i think you merged "NAT Traversal" "Connection Relaying" and "Protocol Muxing" into your node(js)-swarm, but i do not think think this should be merged in the libp2p spec, nor on Go. The go swarm only handles transport multiplexing (connections/streams) to many peers. 

Bringing the protocol multiplexing to something called `host` would be a simple change and essentially one more module. One question, doesn't that mean that swarm doesn't know how to multistream dialed connections into spdy?

> NAT and Relay are separate protocols. In Go, the "Host" handles the protocol multiplexing, and composes (uses) the NAT and Relay protocols.

The way I was pushing forward was that a `transport` implementation is free to be implemented with any feature it needs as long as it respects the interface dial + listen. for example, a libp2p-webrtc could be implemented with a mechanism to do the signalling through the Kademlia Router or some other server, for Swarm, it would only matter that dial(PeerInfo) would dial to that PeerInfo and if some options if preferred, then dial(PeerInfo, {relay: true}) for e.g.

What is the full extent of `Host`?(also happy to call it Node)